### PR TITLE
Add OneHot Transform

### LINKED ISSUE #72: OneHotEncoding
The transform takes categorical columns and converts into multiple columns with 1s and 0s using the CategoricalArrays.jl API.

### PR DESCRIPTION
**Pull request by eliascarv**
This PR implements OneHot transform.

closes #72

### CODE REVIEW

File: src/TableTransforms.jl
Comment (juliohm) [inline src/TableTransforms.jl:41]: I wonder if we could adopt the shorter name OneHot? It should be clear and more concise. What do you think?

File: src/transforms/onehotencoding.jl
Comment (juliohm) [inline src/transforms/onehotencoding.jl:39]: Do we really need to collect the names? Isn't Tables.columnnames always indexable?

File: src/transforms/onehotencoding.jl
Comment (juliohm) [inline src/transforms/onehotencoding.jl:40]: Is this related to the issue we had in other transform where we are avoiding specialization of type in order to revert later? Why do we need the AbstractVector here? Maybe we should use Any as we did in the other case?

File: src/transforms/onehotencoding.jl
Comment (juliohm) [inline src/transforms/onehotencoding.jl:48]: This here should be a new assertion function following our assertion function api. I think we only have assert_continuous at the moment, we should add assert_categorical

File: src/transforms/onehotencoding.jl
Comment (juliohm) [inline src/transforms/onehotencoding.jl:57]: Can you please swap the order here? Why you create a pair with last/first instead of first/last which is more natural? Also, we always think of these pairs as names => cols, not cols => names. Let's try to follow this pattern.

File: src/transforms/onehotencoding.jl
Comment (juliohm) [inline src/transforms/onehotencoding.jl:60]: splice! is a wonderful funciton, I need to use it more!

Comment (juliohm) [reply src/transforms/onehotencoding.jl:60]: Please reorder these lines so that the names => cols pattern is respected.

File: src/transforms/onehotencoding.jl
Comment (juliohm) [inline src/transforms/onehotencoding.jl:72]: Again I wonder why we need to annotate the AbstractVector and if we could just use Any instead

**Review by juliohm — CHANGES_REQUESTED** [review]
Nice usage of splice!, that function is super helpful.

Attached a couple of review suggestions. I think we can improve the code a bit further.

File: src/transforms/onehotencoding.jl
Comment (eliascarv) [reply src/transforms/onehotencoding.jl:39]: Yes, the `splice!` function needs a Vector as the first argument.

Comment (juliohm) [reply src/transforms/onehotencoding.jl:39]: I would call collect or vec in the splice! function then. Calling it here too early makes the reader of the code wonder why this is necessary.

File: src/transforms/onehotencoding.jl
Comment (eliascarv) [reply src/transforms/onehotencoding.jl:40]: So I think Any might be a better choice.

File: src/TableTransforms.jl
Comment (eliascarv) [reply src/TableTransforms.jl:41]: I think that's a good name.

File: src/transforms/onehotencoding.jl
Comment (juliohm) [reply src/transforms/onehotencoding.jl:39]: what about splice!(collect(names), ind, newnms)?

Comment (eliascarv) [reply src/transforms/onehotencoding.jl:39]: The `splice!` function returns the removed value, not the transformed vector. So I need a vector in a variable.

File: src/assertions.jl
Comment (juliohm) [inline src/assertions.jl:13]: We need to use the scitype of the column instead. Check the assert_continuous for an idea. The return of the function will be something like Categorical. Check the type hierarchy of ScientificTypes.jl to learn more.

File: src/transforms/onehot.jl
Comment (juliohm) [inline src/transforms/onehot.jl:32]: So you are reproducing the logic of the ColSpec feature here right? Maybe we can reuse the ColSpec functionality but restrict at construction time what ColSpec are allowed? In this case we want to restrict ourselves to a ColSpec that consist of a single column, right?

File: src/transforms/onehot.jl
Comment (juliohm) [inline src/transforms/onehot.jl:35]: So this function is appending an underscore when the column name already exists right? Can we get rid of it and then inline the implementation inside the apply function? You still have this tendency to create tiny isolated functions that could live in apply :)

File: src/transforms/onehot.jl
Comment (juliohm) [inline src/transforms/onehot.jl:39]:
```suggestion
  names = Tables.columnnames(cols) |> collect
```

**Review by juliohm — CHANGES_REQUESTED** [review]
The PR is looking great already, just a few improvements before we can merge it.

File: src/assertions.jl
Comment (juliohm) [inline src/assertions.jl:14]:
```suggestion
  @assert elscitype(x) <: Finite "The selected column must be categorical."
```

**Review by juliohm — COMMENTED** [review]
Awesome, will merge it soon after the tests pass.

**Review by juliohm — APPROVED** [review]

**Pull request merged**
